---
title: "BCR analysis of ITP samples"
author: "David Schmidt, d.schmidt@sanquin.nl"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F,message = F)
```

```{r libs}
library(tidyverse)
library(Seurat)
library(scRepertoire)
library(future)
library(cowplot)
library(knitr)
library(beepr)
library(magrittr)
library(ggsci)
theme_set(theme_cowplot())

box::use(./box/box)
```

## Load seurat 

major requirement for the attachment is matching contig cell barcodes and barcodes in the row names of the meta data of the seurat or SCE object. If these do not match, the attachment will fail. 

```{r init}
# T cells after clustering
bCells = readRDS("data/processed/20200626 allCells_Clustered.rds")
bCells = subset(bCells,idents=c(6,10))
table(bCells$seurat_clusters) %>% kable(caption="Cells per identified seurat cluster(all samples)")
DimPlot(bCells, reduction = "umap",label = F)
```

## Load contigs

```{r}
# BCR contigs
csv1 <- read.csv("data/contig_files/SAN-DS-G001-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
csv2 <- read.csv("data/contig_files/SAN-DS-G002-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
csv3 <- read.csv("data/contig_files/SAN-DS-G004-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
csv4 <- read.csv("data/contig_files/SAN-DS-G005-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
csv5 <- read.csv("data/contig_files/SAN-DS-G006-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
csv6 <- read.csv("data/contig_files/SAN-DS-G008-BVDJ-all_contig_annotations.csv", stringsAsFactors = F)
contig_list <- list(csv1,csv2,csv3,csv4,csv5,csv6)

# filter non-cells
for (i in seq_along(contig_list)) {
    contig_list[[i]] = contig_list[[i]] %>%  
      filter(is_cell == "True")
    }
```

Half of the contigs of G006 (T269) are not from cells and these should be excluded. Same for G005 (T316).

## BCR contigs

```{r}
combined <- combineBCR(contig_list, 
                       # samples = c("g001", "g002", "g004", "g005", "g006","g008"), 
                       # ID = c("chronic", "transient", "transient", "chronic", "transient", "transient"),
                       samples = c("x", "x", "x", "x", "x","x"), # just a temp hack, fixed in next segment
                       ID = c("T244", "T263", "T281", "T316", "T269","T440")) 

?combineBCR

# delete x_ prefix
names(combined) = gsub("x_","",names(combined))

# cluster info
clusters = data.frame(barcode=rownames(bCells@meta.data),seurat_clusters=bCells@meta.data$seurat_clusters) 
clusters %<>% 
  mutate(seurat_clusters = as_factor(seurat_clusters))

for (i in seq_along(combined)) {
    # get rid of the "sample" 
    combined[[i]]$barcode <- gsub("x_","", x=combined[[i]]$barcode)
    
    # not sure this is necessary
    combined[[i]]$barcode <- gsub("-1","", x=combined[[i]]$barcode)
    
    # filter barcodes
    combined[[i]] = combined[[i]] %>% filter(barcode %in% rownames(bCells@meta.data)) 
    
    # add cluster
    combined[[i]] %<>% 
      left_join(
          clusters, 
          by="barcode"
    )
}



## testing
# bCells@meta.data %>% head
# combined[[1]] %>% head
# (combined[[1]]$barcode %in% rownames(bCells@meta.data)) %>% table
# for (i in seq_along(combined)) print(dim(combined[[i]]))
# names(combined)
# combined[[1]] %>% str
```

### scRep Output

Clonotypes

* "gene+nt" - use the genes comprising the TCR/Ig + the nucleotide sequence of the CDR3 region. `Yassai et al. PMID: 19568742`
* clonotype calling is not incorporating small variations within the CDR3 sequences
* clonotype call is trying to incorporate both loci, i.e, both TCRA and TCRB chains and if a single cell barcode has multiple sequences identified (i.e., 2 TCRA chains expressed in one cell)

**total or relative numbers of unique clonotypes**

```{r}
quantContig(combined, cloneCall="gene+nt", scale = T) + theme(legend.position = "none")
# save_plot(box$figpath("Bcell-unique_Clonotypes.pdf"),last_plot(),base_width=3,base_height=3)

quantContig_output <- quantContig(combined, cloneCall="gene+nt", scale = T, exportTable = T)
quantContig_output %>% kable

# by seurat cluster?
# quantContig(combined, cloneCall="nt", scale = T, group="seurat_clusters") + theme(legend.position = "none")
# save_plot(box$figpath("unique_Clonotypes.pdf"),last_plot(),base_width=3,base_height=3)

```

**relative distribution of clonotypes by abundance** (total number of clonotypes by the number of instances within the sample or run)

```{r}
p = abundanceContig(combined, cloneCall = "gene", scale = F) 
p + box$cornerLegend()
# save_plot(box$figpath("Bcell-abundance_Clonotypes.pdf"),last_plot(),base_width=3,base_height=3)
```


**length distribution of the CDR3 sequences (cloneCall can only be "nt" or "aa")**

the length should reveal a multimodal curve, this is a product of using the NA for the unreturned chain sequence and multiple chains within a single barcode.

```{r}
lengthContig(combined, cloneCall="aa", chains = "combined") 
```

Per chain

```{r}
lengthContig(combined, cloneCall="nt", chains = "single") 
```

**relative usage of vgene**

```{r}
vizVgenes(combined, TCR="TCR1", facet.x = "sample")
save_plot(box$figpath("Bcell-vgenes.pdf"),last_plot(),base_width=6,base_height=3)

```


**clonal space**

we are effectively looking at the relative space occupied by clones at specific proportions

cloneTypes

* Rare = .0001
* Small = .001
* Medium = .01
* Large = .1
* Hyperexpanded = 1

```{r}
clonalHomeostasis(combined, cloneCall = "gene") +
  box$cornerLegend()
# save_plot(box$figpath("BCells-clonalSpace.pdf"),last_plot(),base_width=3,base_height=3)
```

**clonal Proportion**

looking at the relative proportion of the clone to the total, the clonalProportion() function will rank the clones by total number and place them into bins.

split

* 10
* 100
* 1000
* 10000 * 30000 * 100000

```{r}
clonalProportion(combined, cloneCall = "gene") +
  box$cornerLegend()
# save_plot(box$figpath("BCells-clonalProportion.pdf"),last_plot(),base_width=3,base_height=3)
```

**Diversity Analysis**

 Diversity is calculated using four metrics: 1) Shannon, 2) inverse Simpson, 3) Chao1, and 4) Abundance-based Coverage Estimator (ACE). With the former two generally used to estimate baseline diversity and Chao/ACE indices used to estimate the richness of the samples.


```{r}
clonalDiversity(combined, cloneCall = "gene", group = "samples")+
  box$cornerLegend()
# save_plot(box$figpath("bCells-clonalDiversity.pdf"),last_plot(),base_width=4,base_height=3)
```

### Seurat plots

```{r}
# for plotting within seurat
# bCells@meta.data %>% head
# combined[[1]] %>% head
sce = combineExpression(combined, bCells, cloneCall = "gene+nt")
sce@meta.data %>% head
```

The data is there. Trying some working with the data

```{r}
colorblind_vector <- colorRampPalette(c("#FF4B20", "#FFB433", "#C6FDEC", "#7AC5FF", "#0348A6"))
slot(sce, "meta.data")$cloneType <- factor(slot(sce, "meta.data")$cloneType, 
                levels = c("Hyperexpanded (100 < X <= 500)", "Large (20 < X <= 100)", 
                            "Medium (5 < X <= 20)", "Small (1 < X <= 5)", 
                            "Single (0 < X <= 1)", NA))

DimPlot(sce, group.by = "cloneType") +
    scale_color_manual(values = colorblind_vector(5), na.value="grey") +
    box$cornerLegend() 

# save_plot(box$figpath("bCells-clonalExpansion_UMAP.pdf"),last_plot(),base_width=5,base_height=5)

occupiedscRepertoire(sce, x.axis = "cluster") +
    box$cornerLegend()

# save_plot(box$figpath("tCells-clonalExpansion_perCluster.pdf"),last_plot(),base_width=3,base_height=3)

library(circlize)
library(scales)

circles <- getCirclize(sce, groupBy = "cluster")

#Just assigning the normal colors to each cluster
grid.cols <- scales::hue_pal()(length(unique(sce@active.ident)))
names(grid.cols) <- levels(sce@active.ident)

#Graphing the chord diagram
# pdf(file = box$figpath("bCells-Cluster-Clonotype_circle.pdf"), width = 5, height=5)
circlize::chordDiagram(circles, self.link = 1, grid.col = grid.cols)
# dev.off()
```

EXPANDED clones

```{r include=F, eval=F}
sce@meta.data %>% 
  filter(cloneType == "Medium (5 < X <= 20)") %>%
  write.csv2("temp/expandedBcell-Clones.csv")
```



```{r}
# fix some functions of scRepertoire locally
grabMeta <- function(sc) {
    if (inherits(x=sc, what ="Seurat")) {
        meta <- data.frame(sc[[]], slot(sc, "active.ident"))
        colnames(meta)[length(meta)] <- "cluster"
    }
    else if (inherits(x=sc, what ="SummarizedExperiment")){
        meta <- data.frame(colData(sc))
        rownames(meta) <- sc@colData@rownames
        clu <- which(colnames(meta) == "ident")
        colnames(meta)[clu] <- "cluster"
    }
    return(meta)
}

expression2List = function (sc, group)
{
    if (!inherits(x = sc, what = "Seurat") & !inherits(x = sc,
        what = "SummarizedExperiment")) {
        stop("Use a seurat or SCE object to convert into a list")
    }
    meta <- grabMeta(sc)
    unique <- str_sort(as.character(unique(meta[, group])), numeric = TRUE)
    df <- NULL
    for (i in seq_along(unique)) {
        subset <- meta[meta[,group] == unique[i], ] # here;s the fix see https://github.com/ncborcherding/scRepertoire/issues/18
        df[[i]] <- subset
    }
    names(df) <- unique
    return(df)
}
```

```{r}
combined2 <- expression2List(sce,group = "seurat_clusters")
# combined2[[1]] %>% head
clonalDiversity(combined2, cloneCall = "nt") + box$cornerLegend() 
# save_plot(box$figpath("bCells-clusters_diversity.pdf"),last_plot(),base_width=3,base_height=3)

clonalHomeostasis(combined2, cloneCall = "nt") + box$cornerLegend() 
# save_plot(box$figpath("bCells-clusters_clonalHomeostasis.pdf"),last_plot(),base_width=3,base_height=3)

clonalProportion(combined2, cloneCall = "nt") + box$cornerLegend() 
# save_plot(box$figpath("bCells-clusters_clonalProportion.pdf"),last_plot(),base_width=3,base_height=3)

clonalOverlap(combined2, cloneCall="aa", method="overlap") + box$cornerLegend() 
# save_plot(box$figpath("bCells-clusters_clonalOverlap.pdf"),last_plot(),base_width=3,base_height=3)
```


## Outro

```{r outro}
devtools::session_info()
```

